### ISSUE TITLE
Customers losing session

### ISSUE THREAD

**Issue by rafaelpatro** [colinmollenhour/Cm_RedisSession#155]
Hi guys

Occasionally some customers are losing the session. But my local test works great and many customers can buy without problems.

So I don't know how to investigate the problem. Any tip?
Anyone with same problem?

_p.s. Some time ago I've installed that Plumrocket AMP extension and it was great for performance and SEO. My tests with the extension are Ok too. Could it be related with?_

**Comment by colinmollenhour** [comment]
There are all sorts of possible reasons and it is not at all easy to debug.. However, we've applied some fixes for session validation and the "secure cookie" renewal over at OpenMage/magento-lts project so you might consider trying it out.

**Comment by rafaelpatro** [comment]
Yep! Thanks
First reason was my redis limit that achieves 2G. Solved increasing to 4GB. I have so much RAM. How can I set to unlimited?

But I'm facing the same issue below too. And LTS was the best response.
https://magento.stackexchange.com/questions/167616/session-validation-failure-in-magento-1-ee-v-1-14-3-x-and-ce-1-9-3-x

**Comment by rafaelpatro** [comment]
So there is something wrong with my store. Could it grow 1GB in one day?
![image](https://user-images.githubusercontent.com/13813964/46703766-70d5ed80-cbfe-11e8-8a2d-13311fa423b1.png)

**Comment by colinmollenhour** [comment]
If your low limit was causing a lot of key evictions then I would expect to see a spike when the limit is raised.

Note, I strongly recommend running separate redis instances for cache vs session and also FPC vs core cache. With each role having it's own memory limit you reduce risk of one running rampant and causing major issues for the other. E.g. for core cache you want this to never experience evictions and it shouldn't need to since it should be pretty small. FPC and sessions could both use a lot of memory but if they have separate limits then at least they can't negatively impact each other. It is really easy to setup multiple redis instances with Docker. E.g. here is a docker-compose.yml file:

```
version: '3'
services:
  redis-cache:
    image: redis
    ports:
      - "127.0.0.1:6379:6379"
    command: redis-server --maxmemory 512mb --maxmemory-policy volatile-lru --maxmemory-samples 3
    restart: always
    volumes:
      - redis-cache:/data

  redis-session:
    image: redis
    ports:
      - "127.0.0.1:6380:6379"
    command: redis-server --maxmemory 1500mb --maxmemory-policy volatile-lru --maxmemory-samples 3
    restart: always
    volumes:
      - redis-session:/data

  redis-fpc:
    image: redis
    ports:
      - "127.0.0.1:6381:6379"
    command: redis-server --maxmemory 1024mb --maxmemory-policy volatile-lru --maxmemory-samples 3
    restart: always
    volumes:
      - redis-fpc:/data

volumes:
  redis-cache:
  redis-session:
  redis-fpc:
```

Save that to `docker-compose.yml` and then run `docker-compose up` and access each instance at 127.0.0.1 ports 6379, 6380 and 6381.

**Comment by rafaelpatro** [comment]
Thanks @colinmollenhour!
I've followed your tips by applying multiple instances to the databases. It worked partly because the sessions grew less. But it keeps growing insanely.
My last _sessions flush_ was at 2018-10-11. But today (after two weeks) redis reached 3.1G.

**So I'd like to ask whether is something wrong/weird with my setup or 3.1G matches the reality?**

Redis Sessions (uptime 16 days)
  * Size: 3.1G

Google Analytics (from 11 to 26 today)
  * Users: 37658
  * Sessions: 48145

Setup
```xml
        <session_save>db</session_save>
        <redis_session>
            <host>/var/run/redis/sessions.sock</host>
            <port>0</port>
            <password></password>
            <timeout>2.5</timeout>
            <persistent></persistent>
            <db>0</db>
            <compression_threshold>2048</compression_threshold>
            <compression_lib>snappy</compression_lib>
            <log_level>4</log_level>
            <max_concurrency>6</max_concurrency>
            <break_after_frontend>10</break_after_frontend>
            <break_after_adminhtml>30</break_after_adminhtml>
            <fail_after>15</fail_after>
            <first_lifetime>86400</first_lifetime>
            <bot_first_lifetime>7200</bot_first_lifetime>
            <bot_lifetime>21600</bot_lifetime>
            <disable_locking>1</disable_locking>
            <min_lifetime>7200</min_lifetime>
            <max_lifetime>2592000</max_lifetime>
            <log_exceptions>0</log_exceptions>
        </redis_session>
```

**Comment by rafaelpatro** [comment]
Hi @colinmollenhour 
Inspecting sessions today I discovered that _bot_lifetime_ are not working.
So many sessions saved with _googlebot user_agents_. And TTL shows much more time than was configured in _bot_lifetime_.

Should I configure something in Magento in order to recognize user agents?
In last 24hs sessions grows 1.2GB

**Comment by colinmollenhour** [comment]
I tested it stepping through with a debugger and didn't see any issues. Note that your bot lifetime will be bound by the min/max lifetime. The idea behind the *first*  and *bot_first* lifetime is that typically one single page visit does not warrant even having a session. That is, your landing page will probably be a basically static page, either a CMS page or product page or category page. If the user's session is lost after they have only visited one page there is not likely to be any special data in their session that is worth even saving so by reducing the bot_first_lifetime and first_lifetime to something really short like 60 seconds you ensure that 100,000 visits from a bot that doesn't even save your cookie value returned does not waste a bunch of space. By using 7200 seconds for these values you're allowing 2 hours of bot traffic to be wasting your storage space. In my experience a bot does not even really need a session since they aren't doing things like adding products to the cart and you want pages to be indexed as they would be seen by a fresh visitor (e.g. no "recently viewed products") so the only reason not to completely kill their session data immediately is just to be gracious for user agents that may be marked as bots accidentally or in case someone has a site that for some reasons needs the crawler to have a valid session.

**Comment by kanevbg** [comment]
I am observing the same issue in couple projects.  In addition to that the bot_lifetime is not used I see that even the TTL for _first_lifetime_ is not practically working too - in case a website is doing one ajax request on first visit than the visitor will do 2 writes in one actual visitor initiated request thus making ineffective this algorithm. It is better to ignore the write if it is from AJAX request. Another think to ensure is that one requests will not be able to activate the full TTL if for a reason couple or more writes are done from this one request.

**Comment by kanevbg** [comment]
I am providing additional information, I have added logs to getFirstLifetime:
```php
public function getFirstLifetime()
    {
        \Mage::log($this->config, null, 'arebeee.log', true);
        \Mage::log((int) $this->config->descend('first_lifetime'), null, 'arebeee.log', true);
        return (int) $this->config->descend('first_lifetime');
    }
```

Then after request is made the log shows the config object and the raw config value for _first_lifetime_:
![image](https://user-images.githubusercontent.com/11477130/50054484-f61fd600-014a-11e9-984d-f12377eacf5d.png)

When the configuration is accessed from the **_write** method of the adapter, it is holding wrong XML node, it is visible in the above image, here is where the XML node is available in _local.xml_
![image](https://user-images.githubusercontent.com/11477130/50054516-5b73c700-014b-11e9-84a5-7611bb799afb.png)

**Comment by colinmollenhour** [comment]
This could be a bug/behavior in the shutdown procedure whereby some core Magento environment data has already started being destructed. Some ideas to try would be cloning the XML node instead of just assigning a reference to it. However, the calling method (`getLifeTime`) stores the lifetime in a protected property so this must also be getting destructed early? One possible solution would be to store a reference to the handler instance somewhere so it doesn't get destructed so early.. E.g. in the `setSaveHandler` method adding:

```
$GLOBALS['_session_model_instance'] = $this;
```

**Comment by kanevbg** [comment]
Tested with exactly what you suggested, the result is the same. Another strange thing I observe is if I change the constructor of class _Cm_RedisSession_Model_Session_Config_ to
```php
public function __construct($path = 'global/redis_session')
    {
        $this->config = Mage::getConfig()->getNode($path) ?: new Mage_Core_Model_Config_Element('<root></root>');

        \Mage::log('Configuration on construction', null, 'arebeee.log', true);
        \Mage::log($this->config, null, 'arebeee.log', true);
    }
```
(adding the logs) than these logs
```php
public function getFirstLifetime()
    {
        \Mage::log($this->config, null, 'arebeee.log', true);
        \Mage::log((int) $this->config->descend('first_lifetime'), null, 'arebeee.log', true);
        return (int) $this->config->descend('first_lifetime');
    }
```
are not showing up, but I see that the method getFirstLifetime is called still because I have var_dump of its value which is still 0
```php
// Use different lifetime for first write
            if ($lifeTime === null && $this->_sessionWrites <= 1) {
                $firstLifetime = is_null($this->config->getFirstLifetime()) ? self::DEFAULT_FIRST_LIFETIME : $this->config->getFirstLifetime();
                var_dump("first lt: {$firstLifetime}; config FLT: {$this->config->getFirstLifetime()}");
                if ($firstLifetime) {
                    $lifeTime = $firstLifetime * (1+$this->_sessionWrites);
                }
            }
```
which prints `string(26) "first lt: 0; config FLT: 0"`
however the logs are mysteriously missing...

This is the whole log from the request:
```
2018-12-18T18:03:44+00:00 DEBUG (7): Configuration on construction
2018-12-18T18:03:44+00:00 DEBUG (7): Mage_Core_Model_Config_Element Object
(
    [host] => 127.0.0.1
    [port] => 6379
    [password] => Mage_Core_Model_Config_Element Object
        (
        )

    [timeout] => 2.5
    [persistent] => sess-db10
    [db] => 10
    [compression_threshold] => 2048
    [compression_lib] => gzip
    [log_level] => 1
    [max_concurrency] => 6
    [break_after_frontend] => 5
    [break_after_adminhtml] => 30
    [first_lifetime] => 600
    [bot_first_lifetime] => 15
    [bot_lifetime] => 60
    [disable_locking] => 1
    [min_lifetime] => 15
    [max_lifetime] => 2592000
)
```

**Comment by kanevbg** [comment]
Fix is arriving!

**Comment by kanevbg** [comment]
https://github.com/colinmollenhour/Cm_RedisSession/pull/158

**Comment by colinmollenhour** [comment]
Merged #158. Nice work, and thanks for sharing your fix!
